File: src/main/java/cc/mirukuneko/realtrainmodrenewed/rail/util/SwitchType.java
package cc.mirukuneko.realtrainmodrenewed.rail.util;

import net.minecraft.world.level.Level;

import java.util.ArrayList;
import java.util.List;

public abstract class SwitchType {
    public final byte id;
    protected RailMapSwitch[] railMaps = new RailMapSwitch[0];
    protected Point[] points = new Point[0];

    protected SwitchType(int id) {
        this.id = (byte) id;
    }

    public abstract boolean init(List<RailPosition> switchList, List<RailPosition> normalList);

    public void onBlockChanged(Level level) {
    }

    public void onUpdate(Level level) {
        for (Point point : this.points) {
            if (point != null) {
                point.onUpdate(level);
            }
        }
    }

    public RailMapSwitch[] getAllRailMap() {
        return this.railMaps;
    }

    public Point[] getPoints() {
        return this.points;
    }

    public List<RailMap> getOpenRailMaps() {
        if (this.id == 3) {
            return List.of(this.railMaps);
        }
        List<RailMap> open = new ArrayList<>(this.railMaps.length);
        for (RailMapSwitch railMap : this.railMaps) {
            if (railMap != null && railMap.isOpen()) {
                open.add(railMap);
            }
        }
        return open.isEmpty() ? List.of(this.railMaps) : open;
    }

    public int firstOpenRailIndex() {
        if (this.id == 3) {
            return 0;
        }
        for (int i = 0; i < this.railMaps.length; i++) {
            if (this.railMaps[i] != null && this.railMaps[i].isOpen()) {
                return i;
            }
        }
        return 0;
    }

    public int[] getOpenRailIndices() {
        if (this.id == 3) {
            int[] indices = new int[this.railMaps.length];
            for (int i = 0; i < this.railMaps.length; i++) {
                indices[i] = i;
            }
            return indices;
        }
        int count = 0;
        for (RailMapSwitch railMap : this.railMaps) {
            if (railMap != null && railMap.isOpen()) {
                count++;
            }
        }
        if (count == 0) {
            return new int[]{0};
        }
        int[] indices = new int[count];
        int write = 0;
        for (int i = 0; i < this.railMaps.length; i++) {
            if (this.railMaps[i] != null && this.railMaps[i].isOpen()) {
                indices[write++] = i;
            }
        }
        return indices;
    }

    public static class SwitchBasic extends SwitchType {
        public SwitchBasic() {
            super(0);
        }

        @Override
        public boolean init(List<RailPosition> switchList, List<RailPosition> normalList) {
            if (switchList.size() != 1 || normalList.size() != 2) {
                return false;
            }
            RailPosition root = switchList.get(0);
            RailPosition branch1 = normalList.get(0);
            RailPosition branch2 = normalList.get(1);
            RailDir dir = root.getDir(branch1, branch2);
            this.railMaps = new RailMapSwitch[]{
                new RailMapSwitch(root, branch1, dir, RailDir.NONE),
                new RailMapSwitch(root, branch2, dir.invert(), RailDir.NONE)
            };
            this.points = new Point[]{
                new Point(root, railMaps[0], railMaps[1]),
                new Point(branch1, railMaps[0]),
                new Point(branch2, railMaps[1])
            };
            return true;
        }

        @Override
        public void onBlockChanged(Level level) {
            if (level == null || this.railMaps.length < 2) {
                return;
            }
            if (this.railMaps[0].getStartRP().checkRSInput(level)) {
                this.railMaps[0].setState(false);
                this.railMaps[1].setState(true);
            } else {
                this.railMaps[0].setState(true);
                this.railMaps[1].setState(false);
            }
        }
    }

    public static class SwitchSingleCross extends SwitchType {
        public SwitchSingleCross() {
            super(1);
        }

        @Override
        public boolean init(List<RailPosition> switchList, List<RailPosition> normalList) {
            if (switchList.size() != 2 || normalList.size() != 2) {
                return false;
            }
            RailPosition root1 = switchList.get(0);
            RailPosition root2 = switchList.get(1);
            RailDir b0 = RailDir.NONE;
            RailDir b1 = RailDir.NONE;
            RailMapSwitch[] rails = new RailMapSwitch[3];

            RailPosition branch0 = findFirstDifferentDirection(root1, normalList, null);
            RailPosition branch1 = findFirstDifferentDirection(root2, normalList, branch0);
            if (branch0 == null || branch1 == null) {
                return false;
            }

            b0 = root1.getDir(root2, branch0);
            b1 = root2.getDir(root1, branch1);
            rails[0] = new RailMapSwitch(root1, branch0, b0.invert(), RailDir.NONE);
            rails[1] = new RailMapSwitch(root2, branch1, b1.invert(), RailDir.NONE);
            rails[2] = new RailMapSwitch(root1, root2, b0, b1);
            this.railMaps = rails;
            this.points = new Point[]{
                new Point(root1, rails[0], rails[2]),
                new Point(root2, rails[1], rails[2]),
                new Point(rails[0].getStartRP() == root1 ? rails[0].getEndRP() : rails[0].getStartRP(), rails[0]),
                new Point(rails[1].getStartRP() == root2 ? rails[1].getEndRP() : rails[1].getStartRP(), rails[1])
            };
            return true;
        }

        @Override
        public void onBlockChanged(Level level) {
            if (level == null || this.railMaps.length < 3) {
                return;
            }
            if (this.railMaps[2].isGettingPowered(level)) {
                this.railMaps[0].setState(false);
                this.railMaps[1].setState(false);
                this.railMaps[2].setState(true);
            } else {
                this.railMaps[0].setState(true);
                this.railMaps[1].setState(true);
                this.railMaps[2].setState(false);
            }
        }
    }

    public static class SwitchScissorsCross extends SwitchType {
        public SwitchScissorsCross() {
            super(2);
        }

        @Override
        public boolean init(List<RailPosition> switchList, List<RailPosition> normalList) {
            if (switchList.size() != 4) {
                return false;
            }

            RailMapSwitch[] rails = new RailMapSwitch[4];
            RailPosition[][] pairs = new RailPosition[4][2];
            int pairCount = 0;
            for (int i = 0; i < 4; ++i) {
                for (int j = i + 1; j < 4; ++j) {
                    int dirDiff = Math.abs((switchList.get(i).direction & 7) - (switchList.get(j).direction & 7));
                    if (dirDiff > 4) {
                        dirDiff = 8 - dirDiff;
                    }
                    if (dirDiff > 2 && pairCount < 4) {
                        pairs[pairCount++] = new RailPosition[]{switchList.get(i), switchList.get(j)};
                    }
                }
            }
            if (pairCount != 4) {
                return false;
            }

            this.railMaps = new RailMapSwitch[4];
            for (int i = 0; i < 4; ++i) {
                RailDir dir0 = RailDir.NONE;
                RailDir dir1 = RailDir.NONE;
                for (int j = 0; j < 4; ++j) {
                    if (i == j) {
                        continue;
                    }
                    if (pairs[i][0] == pairs[j][0]) {
                        dir0 = pairs[i][0].getDir(pairs[i][1], pairs[j][1]);
                    } else if (pairs[i][0] == pairs[j][1]) {
                        dir0 = pairs[i][0].getDir(pairs[i][1], pairs[j][0]);
                    } else if (pairs[i][1] == pairs[j][0]) {
                        dir1 = pairs[i][1].getDir(pairs[i][0], pairs[j][1]);
                    } else if (pairs[i][1] == pairs[j][1]) {
                        dir1 = pairs[i][1].getDir(pairs[i][0], pairs[j][0]);
                    }
                }
                rails[i] = new RailMapSwitch(pairs[i][0], pairs[i][1], dir0, dir1);
            }
            this.railMaps = rails;

            this.points = new Point[4];
            for (int i = 0; i < 4; ++i) {
                RailPosition rp = switchList.get(i);
                RailMapSwitch rms1 = null;
                RailMapSwitch rms2 = null;
                for (RailMapSwitch railMap : this.railMaps) {
                    if (railMap.getStartRP() == rp || railMap.getEndRP() == rp) {
                        if (rms1 == null) {
                            rms1 = railMap;
                        } else {
                            rms2 = railMap;
                            break;
                        }
                    }
                }
                this.points[i] = new Point(rp, rms1, rms2);
            }
            return true;
        }

        @Override
        public void onBlockChanged(Level level) {
            if (level == null) {
                return;
            }
            RailMapSwitch openRms = null;
            for (int phase = 0; phase < 2; ++phase) {
                for (RailMapSwitch rms : this.railMaps) {
                    if (rms.startDir == rms.endDir) {
                        if (phase == 0) {
                            if (rms.isGettingPowered(level)) {
                                openRms = rms;
                                break;
                            }
                        } else if (rms == openRms) {
                            rms.setState(true);
                        } else {
                            rms.setState(false);
                        }
                    } else if (phase == 1) {
                        rms.setState(openRms == null);
                    }
                }
            }
        }
    }

    public static class SwitchDiamondCross extends SwitchType {
        public SwitchDiamondCross() {
            super(3);
        }

        @Override
        public boolean init(List<RailPosition> switchList, List<RailPosition> normalList) {
            List<RailPosition> all = new ArrayList<>(switchList.size() + normalList.size());
            all.addAll(switchList);
            all.addAll(normalList);
            if (all.size() != 4) {
                return false;
            }
            RailMapSwitch[] rails = new RailMapSwitch[2];
            int k = 0;
            for (int i = 0; i < 4; ++i) {
                for (int j = i + 1; j < 4; ++j) {
                    if (Math.abs((all.get(i).direction & 7) - (all.get(j).direction & 7)) == 4) {
                        rails[k++] = new RailMapSwitch(all.get(i), all.get(j), RailDir.NONE, RailDir.NONE).setState(true);
                        if (k >= 2) {
                            this.railMaps = rails;
                            this.points = new Point[]{
                                new Point(rails[0].getStartRP(), rails[0]),
                                new Point(rails[0].getEndRP(), rails[0]),
                                new Point(rails[1].getStartRP(), rails[1]),
                                new Point(rails[1].getEndRP(), rails[1])
                            };
                            return true;
                        }
                    }
                }
            }
            return false;
        }
    }

    private static RailPosition findFirstDifferentDirection(RailPosition switchMarker, List<RailPosition> normalMarkers, RailPosition used) {
        RailPosition fallback = null;
        for (RailPosition normal : normalMarkers) {
            if (used != null && sameEndpoint(used, normal)) {
                continue;
            }
            if ((switchMarker.direction & 7) != (normal.direction & 7)) {
                return normal;
            }
            if (fallback == null) {
                fallback = normal;
            }
        }
        return fallback;
    }

    private static boolean sameEndpoint(RailPosition a, RailPosition b) {
        return a != null && b != null && a.blockX == b.blockX && a.blockY == b.blockY && a.blockZ == b.blockZ;
    }

}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/rail/util/SwitchType.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.rail.util

import net.minecraft.world.level.Level
import kotlin.math.abs

abstract class SwitchType(@JvmField val id: Byte) {
    @JvmField
    var allRailMap: Array<RailMapSwitch> = emptyArray()
    @JvmField
    var points: Array<Point> = emptyArray()

    abstract fun init(switchList: List<RailPosition>, normalList: List<RailPosition>): Boolean

    open fun onBlockChanged(level: Level?) {}

    open fun onUpdate(level: Level?) {
        for (point in points) {
            point?.onUpdate(level)
        }
    }

    fun getAllRailMap(): Array<RailMapSwitch> = allRailMap

    fun getPoints(): Array<Point> = points

    fun getOpenRailMaps(): List<RailMap> {
        if (id == 3.toByte()) return allRailMap.toList()
        val open = allRailMap.filter { it.isOpen() }
        return open.ifEmpty { allRailMap.toList() }
    }

    fun firstOpenRailIndex(): Int {
        if (id == 3.toByte()) return 0
        for (i in allRailMap.indices) {
            if (allRailMap[i].isOpen()) return i
        }
        return 0
    }

    fun getOpenRailIndices(): IntArray {
        if (id == 3.toByte()) return IntArray(allRailMap.size) { it }
        val indices = allRailMap.indices.filter { allRailMap[it].isOpen() }
        return if (indices.isEmpty()) intArrayOf(0) else indices.toIntArray()
    }

    class SwitchBasic : SwitchType(0) {
        override fun init(switchList: List<RailPosition>, normalList: List<RailPosition>): Boolean {
            if (switchList.size != 1 || normalList.size != 2) return false
            val root = switchList[0]
            val branch1 = normalList[0]
            val branch2 = normalList[1]
            val dir = root.getDir(branch1, branch2)
            allRailMap = arrayOf(
                RailMapSwitch(root, branch1, dir, RailDir.NONE),
                RailMapSwitch(root, branch2, dir.invert(), RailDir.NONE)
            )
            points = arrayOf(
                Point(root, allRailMap[0], allRailMap[1]),
                Point(branch1, allRailMap[0]),
                Point(branch2, allRailMap[1])
            )
            return true
        }

        override fun onBlockChanged(level: Level?) {
            if (level == null || allRailMap.size < 2) return
            if (allRailMap[0].startRP.checkRSInput(level)) {
                allRailMap[0].setState(false)
                allRailMap[1].setState(true)
            } else {
                allRailMap[0].setState(true)
                allRailMap[1].setState(false)
            }
        }
    }

    class SwitchSingleCross : SwitchType(1) {
        override fun init(switchList: List<RailPosition>, normalList: List<RailPosition>): Boolean {
            if (switchList.size != 2 || normalList.size != 2) return false
            val root1 = switchList[0]
            val root2 = switchList[1]
            val branch0 = findFirstDifferentDirection(root1, normalList, null) ?: return false
            val branch1 = findFirstDifferentDirection(root2, normalList, branch0) ?: return false

            val b0 = root1.getDir(root2, branch0)
            val b1 = root2.getDir(root1, branch1)
            val rails = arrayOf(
                RailMapSwitch(root1, branch0, b0.invert(), RailDir.NONE),
                RailMapSwitch(root2, branch1, b1.invert(), RailDir.NONE),
                RailMapSwitch(root1, root2, b0, b1)
            )
            allRailMap = rails
            points = arrayOf(
                Point(root1, rails[0], rails[2]),
                Point(root2, rails[1], rails[2]),
                Point(if (rails[0].startRP == root1) rails[0].endRP else rails[0].startRP, rails[0]),
                Point(if (rails[1].startRP == root2) rails[1].endRP else rails[1].startRP, rails[1])
            )
            return true
        }

        override fun onBlockChanged(level: Level?) {
            if (level == null || allRailMap.size < 3) return
            if (allRailMap[2].isGettingPowered(level)) {
                allRailMap[0].setState(false)
                allRailMap[1].setState(false)
                allRailMap[2].setState(true)
            } else {
                allRailMap[0].setState(true)
                allRailMap[1].setState(true)
                allRailMap[2].setState(false)
            }
        }
    }

    class SwitchScissorsCross : SwitchType(2) {
        override fun init(switchList: List<RailPosition>, normalList: List<RailPosition>): Boolean {
            if (switchList.size != 4) return false
            val rails = arrayOfNulls<RailMapSwitch>(4)
            val pairs = arrayOfNulls<Array<RailPosition>>(4)
            var pairCount = 0
            for (i in 0 until 4) {
                for (j in i + 1 until 4) {
                    var dirDiff = abs((switchList[i].direction.toInt() and 7) - (switchList[j].direction.toInt() and 7))
                    if (dirDiff > 4) dirDiff = 8 - dirDiff
                    if (dirDiff > 2 && pairCount < 4) {
                        pairs[pairCount++] = arrayOf(switchList[i], switchList[j])
                    }
                }
            }
            if (pairCount != 4) return false

            allRailMap = Array(4) { i ->
                var dir0 = RailDir.NONE
                var dir1 = RailDir.NONE
                for (j in 0 until 4) {
                    if (i == j) continue
                    val pi = pairs[i]!!
                    val pj = pairs[j]!!
                    when {
                        pi[0] == pj[0] -> dir0 = pi[0].getDir(pi[1], pj[1])
                        pi[0] == pj[1] -> dir0 = pi[0].getDir(pi[1], pj[0])
                        pi[1] == pj[0] -> dir1 = pi[1].getDir(pi[0], pj[1])
                        pi[1] == pj[1] -> dir1 = pi[1].getDir(pi[0], pj[0])
                    }
                }
                RailMapSwitch(pairs[i]!![0], pairs[i]!![1], dir0, dir1)
            }

            points = Array(4) { i ->
                val rp = switchList[i]
                var rms1: RailMapSwitch? = null
                var rms2: RailMapSwitch? = null
                for (railMap in allRailMap) {
                    if (railMap.startRP == rp || railMap.endRP == rp) {
                        if (rms1 == null) rms1 = railMap
                        else { rms2 = railMap; break }
                    }
                }
                Point(rp, rms1!!, rms2!!)
            }
            return true
        }

        override fun onBlockChanged(level: Level?) {
            if (level == null) return
            var openRms: RailMapSwitch? = null
            for (phase in 0..1) {
                for (rms in allRailMap) {
                    if (rms.startDir == rms.endDir) {
                        when (phase) {
                            0 -> if (rms.isGettingPowered(level)) { openRms = rms; break }
                            1 -> rms.setState(rms == openRms)
                        }
                    } else if (phase == 1) {
                        rms.setState(openRms == null)
                    }
                }
            }
        }
    }

    class SwitchDiamondCross : SwitchType(3) {
        override fun init(switchList: List<RailPosition>, normalList: List<RailPosition>): Boolean {
            val all = mutableListOf<RailPosition>()
            all.addAll(switchList)
            all.addAll(normalList)
            if (all.size != 4) return false
            val rails = arrayOfNulls<RailMapSwitch>(2)
            var k = 0
            for (i in 0 until 4) {
                for (j in i + 1 until 4) {
                    if (abs((all[i].direction.toInt() and 7) - (all[j].direction.toInt() and 7)) == 4) {
                        rails[k++] = RailMapSwitch(all[i], all[j], RailDir.NONE, RailDir.NONE).setState(true)
                        if (k >= 2) {
                            allRailMap = rails.requireNoNulls()
                            points = arrayOf(
                                Point(allRailMap[0].startRP, allRailMap[0]),
                                Point(allRailMap[0].endRP, allRailMap[0]),
                                Point(allRailMap[1].startRP, allRailMap[1]),
                                Point(allRailMap[1].endRP, allRailMap[1])
                            )
                            return true
                        }
                    }
                }
            }
            return false
        }
    }

    companion object {
        private fun findFirstDifferentDirection(
            switchMarker: RailPosition,
            normalMarkers: List<RailPosition>,
            used: RailPosition?
        ): RailPosition? {
            var fallback: RailPosition? = null
            for (normal in normalMarkers) {
                if (used != null && sameEndpoint(used, normal)) continue
                if ((switchMarker.direction.toInt() and 7) != (normal.direction.toInt() and 7)) return normal
                if (fallback == null) fallback = normal
            }
            return fallback
        }

        private fun sameEndpoint(a: RailPosition?, b: RailPosition?): Boolean =
            a != null && b != null && a.blockX == b.blockX && a.blockY == b.blockY && a.blockZ == b.blockZ
    }
}
